providers/runtime.go

```diff
@@ -216,6 +224,32 @@ func (r *Runtime) Connect(req *plugin.ConnectReq) error {
 	if err != nil {
 		return err
 	}
+
+	// TODO: This is a stopgap that detects if the connect call returned an asset
+	// that is different from the provider we used for connecting. We will keep
+	// supporting this approach throughout v9 but plan to change it in the future,
+	// so that the connect call sticks to connecting only and instead introduce
+	// a separate discover call to handle this behavior.
+	//
+	// This stopgap makes sure that if the connection indicates a different provider,
+	// it is the intention of the provider author to switch the asset to said provider.
+	//
+	// Additionally, we do not loop this connect+recheck approach indefinitely.
+	// We only run it once and only accept one asset switch. This will be
+	// changed once we have an explicit discover call in plugins.
+	postProvider, err := r.providerForAsset(r.Provider.Connection.Asset)
+	if err != nil {
+		return err
+	}
+	if postProvider.ID != r.Provider.Instance.ID {
+		req.Asset = r.Provider.Connection.Asset
```

Comment:
@arlimus I needed to add this line to make it also work with plugins. With this change, the results also looks like the one in v8.

I tried to set all this inside the GCP provider, but it does not get transferred to the os provider.

Reply:
Makes sense, thank you for the catch!